### PR TITLE
Do not use MaxTimestamp in a get in MVCCConditionalPut

### PR DESCRIPTION
**Pull request by kkaneda**
The behavior looked a bit weird to me, but I wasn't sure what's the right behavior here..

### CODE REVIEW

File: storage/engine/mvcc_test.go [hunk 1]
```diff
@@ -1236,6 +1236,43 @@ func TestMVCCResolveTxn(t *testing.T) {
...
+	err := MVCCPut(engine, nil, testKey1, makeTS(1, 0), value1, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = MVCCPut(engine, nil, testKey1, makeTS(3, 0), value2, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// QUESTION(kkaneda): Should this fail with WriteTooOldError?
```
Comment (bdarnell) [inline storage/engine/mvcc_test.go]: Yeah, these errors look backwards. If you decomposed the ConditionalPut into a transaction with a Get and Put, the first would match the existing value and then fail with a WriteTooOld error, while the second would look for the value at t=2 and get ConditionFailedError. (Actually, the timestamps in this test should result in ReadWithinUncertaintyInterval, but with more space between them you'd still get WriteTooOld and ConditionFailed).

Comment (kkaneda) [inline storage/engine/mvcc_test.go]: Thanks for taking a look, @bdarnell ! Ah, the explanation with Get/Put transaction makes sense. 

Is it OK to simply stop using `proto.MaxTimestamp` in `MVCCConditionalPut`? I don't fully understand the comment, but it seems we anyway get a `WriteTooOld` error if there is a write intent with a newer timestamp.

Comment (bdarnell) [inline storage/engine/mvcc_test.go]: I think that's fine, but we should confirm with @spencerkimball.

File: storage/engine/mvcc_test.go [hunk 2]
```diff
@@ -1236,6 +1236,39 @@ func TestMVCCResolveTxn(t *testing.T) {
 	}
 }
 
+// TestMVCCConditionalPutOldTimestamp tests a case where a conditional
+// put with an older timestamp happens after a put with a newer timestamp.
```
Comment (tschottdorf) [inline storage/engine/mvcc_test.go]: explain what the test is for: That the conditional put uses the actual value at the timestamp as the basis for comparison first, and then may fail later with a WriteIntentError if that timestamp isn't recent.

File: storage/engine/mvcc.go [hunk 3]
```diff
@@ -835,11 +835,7 @@ func MVCCIncrement(engine Engine, ms *MVCCStats, key proto.Key, timestamp proto.
 // containing the actual value.
 func MVCCConditionalPut(engine Engine, ms *MVCCStats, key proto.Key, timestamp proto.Timestamp, value proto.Value,
 	expValue *proto.Value, txn *proto.Transaction) error {
-	// Handle check for non-existence of key. In order to detect
-	// the potential write intent by another concurrent transaction
-	// with a newer timestamp, we need to use the max timestamp
-	// while reading.
-	existVal, _, err := MVCCGet(engine, key, proto.MaxTimestamp, true /* consistent */, txn)
+	existVal, _, err := MVCCGet(engine, key, timestamp, true /* consistent */, txn)
```
Comment (tschottdorf) [inline storage/engine/mvcc.go]: Can you add a comment here explaining that it's ok to read at `timestamp`? A naive look would suspect that there might be a bug here.

**Comment by tschottdorf** [comment]
LGTM mod @spencerkimball

File: storage/engine/mvcc_test.go (hunk 2 above)
Comment (kkaneda) [inline storage/engine/mvcc_test.go]: Sure! Updated.

File: storage/engine/mvcc.go (hunk 3 above)
Comment (kkaneda) [inline storage/engine/mvcc.go]: Sure, added the following comment:

```
	// Use the specified timestamp to read the value. When a write
	// with newer timestamp exists, either of the following will
	// happen:
	// - If the actual value is equal to the expected value,
	//   a WriteTooOldError error is returned from MVCCPut.
	// - If the actual value is not equal to the expected value,
	//   ConditionFailedError is returned.
```

**Comment by yananzhi** [comment]
@kkaneda Use MaxTimestamp in a get in MVCCConditionalPut is helpful to check an unique constraint.

**Comment by kkaneda** [comment]
Hmm, @yananzhi , could you elaborate? I'm guessing `MaxTimestamp` is useful to check the existence of keys regardless of their values.

**Comment by yananzhi** [comment]
@kkaneda  According to my understanding, a insert statement need be converted to several condition-put KV command if the target table has primary key. the condition put will check the existence of primary key by read the key with MaxTimestamp. But I check the code just now, the insert statement was converted to KV put for each column, It seems don't consider the unique constraint, If the table  exist the same primary key, KV put will overwrite the key rather than report a violation unique constraint error.
    Assume a insert statement was converted condition-put KV command, the expected value is set to nil to check the existence of primary key. If the key there has a newest commit version with a higher timestamp than current command's timestamp, and the newest commit version was generate by a delete KV command. If we use the current command's timestamp to do the MVCCGet, we will get a value which was not equal to nil and the condition put command return a ConditionFailedError which was not a retryable error, and the txn will abort. If we use the MaxTimestamp to do the condition put, we will get a nil value for MVCCGet, and the do the MVCCPut, MVCCPut wil return a WriteTooOldError which is a retryable error, the condtion put will retry in store level, the condition may succeed in subsequent retry.

**Comment by kkaneda** [comment]
Thanks for the detailed explanation, @yananzhi ! 

If I understand correctly, the following is the scenario you mentioned:

1) Insert key K with value V at timestamp T.
2) Delete key K at timestamp T' (> T).
3) Insert key K with value V' at timestamp T'' (T < T'' < T'). The insert returns `ConditionFailedError`, not `WriteTooOldError`. The transaction is not automatically retried.

IMO, this is the expected behavior. If we look at the timestamps of the events. The deletion happens after the second insertion. For example, if Step 3 happens before Step 2, the second insertion should return `ConditionFailedError`:

1) Insert key K with value V at timestamp T.
2) Insert key K with value V' at timestamp T' (T' > T). The insert returns `ConditionFailedError`.
3) Delete key K at timestamp T'' (> T').

I don't fully understand how KV clients work, so I might be missing your point..

File: storage/engine/mvcc.go [hunk 4]
```diff
@@ -835,11 +835,14 @@ func MVCCIncrement(engine Engine, ms *MVCCStats, key proto.Key, timestamp proto.
...
 	expValue *proto.Value, txn *proto.Transaction) error {
-	// Handle check for non-existence of key. In order to detect
-	// the potential write intent by another concurrent transaction
-	// with a newer timestamp, we need to use the max timestamp
-	// while reading.
-	existVal, _, err := MVCCGet(engine, key, proto.MaxTimestamp, true /* consistent */, txn)
+	// Use the specified timestamp to read the value. When a write
+	// with newer timestamp exists, either of the following will
+	// happen:
+	// - If the actual value is equal to the expected value,
```
Comment (spencerkimball) [inline storage/engine/mvcc.go]: This comment isn't quite right.

    // - If the conditional check succeeds, either a WriteTooOldError or WriteIntentError
    //   is return, depending on whether the newer write is an intent.
    // - If the conditional check fails, a ConditionFailedError is returned.

**Comment by spencerkimball** [comment]
This seems like a slightly more correct interpretation. You might add to the comments in MVCCConditionalPut to be explicit about the interpretation of the conditional clause according to the operation timestamp.

LGTM

File: storage/engine/mvcc.go (hunk 4 above)
Comment (kkaneda) [inline storage/engine/mvcc.go]: Ah, I see. Updated.

**Comment by kkaneda** [comment]
Thanks for the review, @spencerkimball ! Added the following comment to the method description:

```
// The condition check reads a value from the key using the same operational
// timestamp as we use to write a value.
```

@yananzhi , please let me know if you still see an issue with this change.

**Comment by yananzhi** [comment]
@kkaneda  Just ignore my opinion, Like you, I really don't know what is expected behavior for this interface and what scenario this interface can be used.

**Comment by spencerkimball** [comment]
LGTM

**Pull request merged**